get_element_cpo/P2769.bs

```diff
@@ -460,119 +467,75 @@ concept /*tuple-like*/ = !std::is_reference_v<T> &&
-the current `std::ranges::get` function is used with explicit template arguments.
+```cpp
+namespace my {
+    struct type{};
 
-Please see the example that explains the idea and shows how it might look like. A full implementation with examples
-is available [here](https://godbolt.org/z/boKcMWd5f).
+    template <std::size_t I, typename... Args>
+    decltype(auto) get(std::tuple<Args...>& t)
+    {
+        return std::get<I>(t);
+    }
 
-```cpp
-namespace std
-{
-namespace ranges
-{
-// Necessary to make namespace __detail being considered by ADL
-// for get<0>(std::ranges::subrange<something>{}) without moving
-// the subrange itself to another namespace
-namespace __detail
-{
-struct adl_hook {};
 }
 
-// thanks to the empty-base optimization, inheriting adl_hook does not break ABI
-template <class T> class subrange : __detail::adl_hook {
-    public: T whatever;
-};
-
-namespace __detail
+int main()
 {
-template <std::size_t, class T>
-auto get(subrange<T> x) {
-    return x.whatever;
+    std::tuple<my::type> t;
+
+    // std::get_element<0>(t); // call to ADL get is ambiguous here
+
+    std::apply([](auto) { /* do something */ }, t); // potential breaking change, apply uses get_element
 }
-} // namespace __detail
-} // namespace ranges
+```
 
-using std::ranges::__detail::get;
-} // namespace std
+This breaking change could be avoided by checking in the `get_element` implementation if the function argument type is one
+five types `std::get` can work with. The implementation with pseudo-code would look like:
 
+
+```cpp
 namespace std
 {
-namespace ranges
-{
 namespace __detail
 {
-// Introduce Args... to cover the case of calling get with explicit template arguments
-template <std::size_t _Ip, typename... Args>
-struct __get_fn
+template <std::size_t _Ip>
+struct __get_element_fn
 {
-    // No more than std::tuple_size_v template arguments should be allowed
-    template <typename _TupleLike> requires (sizeof...(Args) <= std::tuple_size_v<std::remove_cvref_t<_TupleLike>>
-                                     && __are_tuple_elements_convertible_to_args<std::remove_cvref_t<_TupleLike>, Args...>::value)
-    decltype(auto) operator()(_TupleLike&& __tuple_like) const
+    template <typename _TupleLike>
+    auto operator()(_TupleLike&& __tuple_like) const ->
+         decltype(auto)
     {
+        constexpr bool __get_member_well_formed = requires {
+            std::forward<_TupleLike>(__tuple_like).template get<_Ip>();
+        };
+        // using __type = std::remove_cvref_t<_TupleLike>'
+        // if constexpr (type is std::tuple or std::pair or std::array or std::ranges::subrange or std::complex)
+        // {
+        //     std::get<_Ip>(std::forward<_TupleLike>(__tuple_like));
+        // }
+        if constexpr (__get_member_well_formed)
+        {
+            return std::forward<_TupleLike>(__tuple_like).template get<_Ip>();
+        }
```

Comment:
I think it's interesting to point out that people seemingly do unqualified call to std::get everywhere even if they are not supposed to https://github.com/search?q=%2F%5B%5E.%3A%5Dget%3C%5Cd%2B%3E%5C(%2F%20language%3AC%2B%2B&type=code

Anyone that adds a specialization can already get inconsistent results